geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSet.java

```diff
@@ -173,20 +176,23 @@ public int getDSFID() {
   }
 
   protected synchronized byte[] memberAdd(byte[] memberToAdd, byte[] scoreToAdd) {
-    byte[] oldScore = null;
-
-    OrderedSetEntry newEntry = new OrderedSetEntry(memberToAdd, scoreToAdd);
-    OrderedSetEntry orderedSetEntry = members.put(memberToAdd, newEntry);
-    if (orderedSetEntry == null) {
+    OrderedSetEntry existingEntry = members.get(memberToAdd);
+    if (existingEntry == null) {
+      OrderedSetEntry newEntry = new OrderedSetEntry(memberToAdd, scoreToAdd);
+      members.put(memberToAdd, newEntry);
       scoreSet.add(newEntry);
-      sizeInBytes += calculateSizeOfFieldValuePair(memberToAdd, scoreToAdd);
+      // Without this adjustment, we count the entry and member name array twice, since references
+      // to them appear in both backing collections.
+      sizeInBytesAdjustment += newEntry.getSizeInBytes() + calculateByteArraySize(memberToAdd);
+      return null;
     } else {
-      scoreSet.remove(orderedSetEntry);
-      scoreSet.add(newEntry);
-      oldScore = orderedSetEntry.getScoreBytes();
-      sizeInBytes += scoreToAdd.length - oldScore.length;
+      scoreSet.remove(existingEntry);
+      byte[] oldScore = existingEntry.scoreBytes;
+      existingEntry.updateScore(stripTrailingZeroFromDouble(scoreToAdd));
+      members.put(memberToAdd, existingEntry);
```

Comment:
why is this members.put needed when we modified an existing entry? Is it simply to update the size if the byte array form has a different length? We are using extra memory by storing both the byte array form and the double form. We would use less memory if we just stored it as a "double". In this case storing it in both forms uses more CPU since we need to do the extra put. What operations are being optimized by storing it in both forms?

Reply:
I think that the reason we store both the byte array and double is that some of the SortedSet commands (zadd and zrem) were implemented before we had an actual sorted set implementation, so were just using the member and score byte arrays in a `SizeableObject2ObjectOpenCustomHashMapWithCursor`. With that original implementation, it was faster to leave the scores in their byte array format at all times, but after new commands and the `OrderStatisticsSet` were added to RedisSortedSet, this is no longer the case.

I think that the only time we need a byte array rather than a double is when we're sending the response, so we can eliminate the field from `AbstractOrderedStatisticsEntry` entirely and just change how we return from certain sorted set methods. However, this is a potentially large refactor that's not strictly within the scope of this PR, so I'd prefer to open a new ticket for it rather than do it as part of this one.

Reply:
sounds good

Reply:
you might want to add a note to that new ticket that this extra members.put call can be removed once this modify no longer changes the size because we are just updating the score stored as a "double".

Reply:
Will do

Reply:
Filed GEODE-9446 to track the refactoring work.